pkg/etcd/etcdv3.go

```diff
@@ -144,3 +136,117 @@ func (b V3Backup) getTmpDir() string {
 
 	return *b.tmpDir
 }
+
+func (b V3Backup) compactAndDefrag() error {
+	b.Logger.Debugf(context.Background(), "Compacting etcd instance")
+	// Get latest revision.
+	output, err := b.runEtcdctlCmd([]string{
+		"endpoint",
+		"status",
+		`--write-out=json`,
+	})
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	revision, err := getRevision(output)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	b.Logger.Debugf(context.Background(), "Revision is %d", revision)
+
+	_, err = b.runEtcdctlCmd([]string{
+		"compact",
+		fmt.Sprintf("%d", revision),
+	})
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	b.Logger.Debugf(context.Background(), "Compacted etcd instance")
+
+	b.Logger.Debugf(context.Background(), "Defragging etcd instance")
+
+	_, err = b.runEtcdctlCmd([]string{
+		"defrag",
+		"--command-timeout=60s",
+		"--dial-timeout=60s",
+	})
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	b.Logger.Debugf(context.Background(), "Defragged etcd instance")
+
+	return nil
+}
+
+func (b V3Backup) runEtcdctlCmd(etcdctlArgs []string) ([]byte, error) {
+	etcdctlEnvs := []string{"ETCDCTL_API=3"}
+
+	if b.Endpoints != "" {
+		etcdctlArgs = append(etcdctlArgs, "--endpoints", b.Endpoints)
+	}
+	if b.CACert != "" {
+		etcdctlArgs = append(etcdctlArgs, "--cacert", b.CACert)
+	}
+	if b.Cert != "" {
+		etcdctlArgs = append(etcdctlArgs, "--cert", b.Cert)
+	}
+	if b.Key != "" {
+		etcdctlArgs = append(etcdctlArgs, "--key", b.Key)
+	}
+
+	log, err := exec.Cmd(key.EtcdctlCmd, etcdctlArgs, etcdctlEnvs, b.Logger)
+	if err != nil {
+		return nil, errors.New(string(log))
+	}
+
+	return log, nil
+}
+
+func getRevision(output []byte) (int32, error) {
+	type endpointStatusOutputStatusHeader struct {
+		Revision int32
+	}
+
+	type endpointStatusOutputStatus struct {
+		Header endpointStatusOutputStatusHeader
+	}
+
+	type endpointStatusOutput struct {
+		Status endpointStatusOutputStatus
+	}
+
+	// [
```

Comment:
???

Reply:
that is the JSON I am about to parse after the comment

Reply:
Do we need the comment left there? There's also an example of it in the test if needed.

Reply:
okay dropping